R/SailaniIPOPData.R

```diff
@@ -0,0 +1,95 @@
+#' Obtain the Sailani deep longitudinal multiomics data
+#'
+#' Obtain the deep longitudinal multiomics data from Sailani et al. (2020).
+#'
+#' @details
+#' The SailaniIPOP dataset contains longitudinal multiomics data from profiling 
+#' of 105 generally healthy individuals. This dataset consists of 16S rRNA gene 
+#' based profiling of gut and nasal microbiota. Host transcriptome, proteome and 
+#' metabolome from plasma, cytokine profile, and associated clinical data. The
+#' samples were collected over a period of 4 years. The data set can be used 
+#' to investigate longitudinal dynamics and interaction between microbiota and 
+#' host associated factors.  
+#'
+#' This microbiota data as currently available as relative abundances. Hierarchical
+#' taxonomic information is unavailable for microbiota data.
+#'
+#' Column metadata includes the time points, sex, subject identifier, sample
+#' identifier and treatment group.
+#'
+#'
+#' All data are downloaded from ExperimentHub and cached for local re-use.
+#'
+#' @return for  \code{SailaniIPOPData} a \linkS4class{MultiAssayExperiment} object
+#'   consisting of two \linkS4class{TreeSummarizedExperiment} objects and 
+#'   five \linkS4class{SummarizedExperiment} objects.
+#'
+#' @author Sudarshan Shetty and Felix G.M. Ernst
+#'
+#' @references
+#' Sailani, M.R., Metwally, A.A., Zhou, W. et al. (2020) Deep longitudinal multiomics 
+#' profiling reveals two biological seasonal patterns in California.
+#' \emph{Nat Commun.} 11, 4933 \url{https://doi.org/10.1038/s41467-020-18758-1}
+#'
+#' Zhou, W., Sailani, M.R., Contrepois, K. et al. (2019) Longitudinal multi-omics of 
+#' host-microbe dynamics in prediabetes.
+#' \emph{Nature} 569, 663-671 \url{https://doi.org/10.1038/s41586-019-1236-x}
+#'
+#' @name SailaniIPOPData
+#' @export
+#'
+#' @examples
+#' mae <- SailaniIPOPData()
+#' #tse <- SailaniIPOPData()
+SailaniIPOPData <- function() {
```

Comment:
Without having looked at the data: Can you process the samples in away, that the common colData elements are stored in the colData of the `MAE` instead of the individual `experiments`?

Reply:
Its a challenge because, I had to use custom rownames e.g. gutbiota_1, gutbiota_2, etc. Cannot find linkages between samples from different assays. Cannot use SubjectIds because these are longitudinal samples (n>1 for each subject). Subject-Time combination is issue because gutMicrobiota colData has no time information.

Reply:
I am referring the `colData` not the `rowData`, Since a sampleMap is not included, I would imagine it is a 1:1 mapping, isn't it?

Reply:
There are different row numbers and sample numbers.
```
 [1] gutMicrobiota: TreeSummarizedExperiment with 96 rows and 855 columns
 [2] nasalMicrobiota: TreeSummarizedExperiment with 80 rows and 837 columns
 [3] hostTranscriptome: SummarizedExperiment with 10339 rows and 882 columns
 [4] hostProteome: SummarizedExperiment with 302 rows and 950 columns
 [5] hostMetabolome: SummarizedExperiment with 724 rows and 981 columns
 [6] hostCytokines: SummarizedExperiment with 66 rows and 964 columns
 [7] hostClinical: SummarizedExperiment with 51 rows and 969 columns

```

Reply:
hmm.

For this to make sense, a `sampleMap` needs to be generated. Who does the `sampleMap` look like currently?
It is generated on the fly I guess. From this mapping the common `colData` variables should be extractable, shouldn't they?

Reply:
It is generated automatically. 
```
> sampleMap(mae)
DataFrame with 6438 rows and 3 columns
             assay      primary      colname
          <factor>  <character>  <character>
1    gutMicrobiota   gutbiota_1   gutbiota_1
2    gutMicrobiota   gutbiota_2   gutbiota_2
3    gutMicrobiota   gutbiota_3   gutbiota_3
4    gutMicrobiota   gutbiota_4   gutbiota_4
5    gutMicrobiota   gutbiota_5   gutbiota_5
...            ...          ...          ...
6434  hostClinical clinical_965 clinical_965
6435  hostClinical clinical_966 clinical_966
6436  hostClinical clinical_967 clinical_967
6437  hostClinical clinical_968 clinical_968
6438  hostClinical clinical_969 clinical_969
```
As far as I can see there is not 1:1 mapping.
`primary` and `colname` are same essentially colnames in `assays`. 

```
head(assay(mae[["gutMicrobiota"]]))[1:3,1:5]
# otuput
                       gutbiota_1  gutbiota_2  gutbiota_3 gutbiota_4  gutbiota_5
phylum_Actinobacteria 0.000449469 0.000175673 0.000597467  0.0000575 0.001112673
phylum_Bacteroidetes  0.650866093 0.730517883 0.178363738  0.7407258 0.554579527
phylum_Firmicutes     0.225080386 0.184807814 0.796861308  0.2265165 0.364195362
```

Reply:
Ah ok, so no mapping then.

So if you can reconstruct which column belongs to which sample for each of the experiments and add names like `sample-1`, `sample-2` and so forth, a mapping of samples across the experiments should be generated automatically

Reply:
For example, the output above could look like this for a non-automatic mapping:

```
> sampleMap(mae)
DataFrame with 6438 rows and 3 columns
             assay      primary      colname
          <factor>  <character>  <character>
1    gutMicrobiota   sample_1   gutbiota_1
2    gutMicrobiota   sample_2   gutbiota_2
3    gutMicrobiota   sample_3   gutbiota_3
4    gutMicrobiota   sample_4   gutbiota_4
5    gutMicrobiota   sample_5   gutbiota_5
...            ...          ...          ...
6434  hostClinical sample_965 clinical_965
6435  hostClinical sample_966 clinical_966
6436  hostClinical sample_967 clinical_967
6437  hostClinical sample_968 clinical_968
6438  hostClinical sample_969 clinical_969
```

The link is then established through the relation of primary to colname. 

In case of automatic mapping it should look like this:
```
> sampleMap(mae)
DataFrame with 6438 rows and 3 columns
             assay      primary      colname
          <factor>  <character>  <character>
1    gutMicrobiota   sample_1   sample_1
2    gutMicrobiota   sample_2   sample_2
3    gutMicrobiota   sample_3   sample_3
4    gutMicrobiota   sample_4   sample_4
5    gutMicrobiota   sample_5   sample_5
...            ...          ...          ...
6434  hostClinical sample_965 sample_965
6435  hostClinical sample_966 sample_966
6436  hostClinical sample_967 sample_967
6437  hostClinical sample_968 sample_968
6438  hostClinical sample_969 sample_969
```

Reply:
But we don't know if `sample-1` in gutMicrobiota is same as `sample-1` from nasalMicrobiota. They could be different. e.g. `subject1-time1-sample1` and `subject1-time2-sample1`. My thinking is that the sample should come from same subject at same time point for different measurements for it to be a linked -omics assay. Right?

Reply:
I think this should be reconstructable from the original data. What where the original column names?

Reply:
Examples:
GUT MICROBIOTA
```
> gut_df_Data[1:5,1:5]
  phylum_Actinobacteria phylum_Bacteroidetes phylum_Firmicutes phylum_Proteobacteria phylum_Verrucomicrobia
1           0.000449469            0.6508661         0.2250804           0.007364381            0.015835148
2           0.000175673            0.7305179         0.1848078           0.003899937            0.008186354
3           0.000597467            0.1783637         0.7968613           0.000358480            0.002788178
4           0.000057500            0.7407258         0.2265165           0.005976840            0.001206862
5           0.001112673            0.5545795         0.3641954           0.038357929            0.000058600
```

```
> Gut_annotation_colData[1:5,1:5]
   SubjectID Date      IRIS   BMI Time
V2   ZOZOW1T <NA>  5-Dec-14 59.48   NA
V3   ZOZOW1T <NA>  7-Dec-14 59.48   NA
V4   ZOZOW1T <NA>  9-Dec-14 59.48   NA
V5   ZOZOW1T <NA>  5-Jan-15 59.48   NA
V6   ZOZOW1T <NA> 17-Jun-15 59.48   NA
```
NASAL MICROBIOTA
```
> Nasal_df_Data[1:5,1:5]
  phylum_Actinobacteria phylum_Bacteroidetes phylum_Cyanobacteria.Chloroplast phylum_Firmicutes
1            0.06080447          0.000822571                      0.000493543         0.1060459
2            0.33716620          0.006809860                      0.003706350         0.4417701
3            0.60363358          0.000525337                      0.000000000         0.1946372
4            0.77129388          0.000000000                      0.000000000         0.1851005
5            0.86914848          0.004392624                      0.000428831         0.1160742
  phylum_Fusobacteria
1         0.000000000
2         0.000357239
3         0.000000000
4         0.000000000
5         0.000000000
```
```
> Nasal_annotation_colData[1:5,1:5]
   SubjectID                Date IRIS   BMI Time
V2   ZOZOW1T 0014-12-09 08:46:26   IS 21.47  343
V3   ZOZOW1T 0015-06-17 08:46:26   IS 21.47  168
V4   ZOZOW1T 0015-06-19 08:46:26   IS 21.47  170
V5   ZOZOW1T 0016-07-05 08:46:26   IS 21.47  187
V6   ZOZOW1T 0016-07-08 08:46:26   IS 21.47  190
```

Reply:
So the map should be generated via the Subject_ID of the colData.

The assay data is clearly transposed.

It think the following steps should make sense:
1. transpose the assay data, so that `phylum_Actinobacteria` is the rowname
2. Created taxonomic data from these rownames and populate the `Phylum` column of the `rowData`
3. Create colnames reflecting the subject ID, perhaps by using `make.unique(SubjectID)`
4. Create `colData` for the `MAE` for each unique subject. This should include `SubjectID`, `BMI` and other variables which are not changing. Maybe the `BMI` changes in the measurements but I am not sure.
5. Create a `sampleMap`, which maps the `colData` of the `MAE` to each `colData` of the `experiments`

PS: The IRIS column seams to be the date in `Gut_annotation_colData`. Was this misimportated?

Reply:
in the commit, I transposed in the assay. 
Thanks for pointing this out `Created taxonomic data from these rownames and p...` now i see something we need to resolve. 
```
> colnames(Nasal_df_Data)
 [1] "phylum_Actinobacteria"                  "phylum_Bacteroidetes"                  
 [3] "phylum_Cyanobacteria.Chloroplast"       "phylum_Firmicutes"                     
 [5] "phylum_Fusobacteria"                    "phylum_Proteobacteria"                 
 [7] "phylum_unclassified_Bacteria"           "class_Actinobacteria"                  
 [9] "class_Alphaproteobacteria"              "class_Bacilli"                         
[11] "class_Bacteroidia"                      "class_Betaproteobacteria"              
[13] "class_Chloroplast"                      "class_Clostridia"                      
[15] "class_Flavobacteria"                    "class_Fusobacteria"                    
[17] "class_Gammaproteobacteria"              "class_Negativicutes"                   
[19] "class_unclassified_Bacteria"            "order_Actinomycetales"                 
[21] "order_Bacillales"                       "order_Bacteroidales"                   
[23] "order_Burkholderiales"                  "order_Clostridiales"                   
[25] "order_Enterobacteriales"                "order_Flavobacteriales"                
[27] "order_Fusobacteriales"                  "order_Lactobacillales"                 
[29] "order_Neisseriales"                     "order_Pasteurellales"                  
[31] "order_Pseudomonadales"                  "order_Rhizobiales"                     
[33] "order_Selenomonadales"                  "order_unclassified_Bacteria"           
[35] "order_unclassified_Chloroplast"         "family_Actinomycetaceae"               
[37] "family_Bacillaceae.1"                   "family_Bacillales_Incertae.Sedis.XI"   
[39] "family_Carnobacteriaceae"               "family_Chloroplast"                    
[41] "family_Clostridiales_Incertae.Sedis.XI" "family_Comamonadaceae"                 
[43] "family_Corynebacteriaceae"              "family_Enterobacteriaceae"             
[45] "family_Flavobacteriaceae"               "family_Fusobacteriaceae"               
[47] "family_Lachnospiraceae"                 "family_Micrococcaceae"                 
[49] "family_Moraxellaceae"                   "family_Neisseriaceae"                  
[51] "family_Pasteurellaceae"                 "family_Porphyromonadaceae"             
[53] "family_Prevotellaceae"                  "family_Propionibacteriaceae"           
[55] "family_Staphylococcaceae"               "family_Streptococcaceae"               
[57] "family_Veillonellaceae"                 "family_unclassified_Actinomycetales"   
[59] "family_unclassified_Bacteria"           "genus_Actinomyces"                     
[61] "genus_Anaerococcus"                     "genus_Corynebacterium"                 
[63] "genus_Dolosigranulum"                   "genus_Finegoldia"                      
[65] "genus_Fusobacterium"                    "genus_Gemella"                         
[67] "genus_Granulicatella"                   "genus_Neisseria"                       
[69] "genus_Peptoniphilus"                    "genus_Porphyromonas"                   
[71] "genus_Prevotella"                       "genus_Propionibacterium"               
[73] "genus_Rothia"                           "genus_Staphylococcus"                  
[75] "genus_Streptococcus"                    "genus_Streptophyta"                    
[77] "genus_Veillonella"                      "genus_unclassified_Actinomycetales"    
[79] "genus_unclassified_Bacteria"            "genus_unclassified_Neisseriaceae" 
```
Not aware how to convert this to taxonomy file with a code. Manually check taxonomies might be too much...

Reply:
I think the best way, would be to populate each taxonomic level as a column and set the missing values to `NA`. Introduce another column called `taxonomicLevel`, where you save the level used. This way the result should look like the result you get from using `unsplitByRanks(splitByRanks())`